src/ripple/shamap/impl/SHAMapInnerNode.cpp

```diff
@@ -132,35 +132,86 @@ SHAMapInnerNode::makeFullInner(
     if (data.size() != branchFactor * uint256::bytes)
         Throw<std::runtime_error>("Invalid FI node");
 
-    auto ret = std::make_shared<SHAMapInnerNode>(0, branchFactor);
-
     SerialIter si(data);
 
+    // Determine the non-empty branches so we only allocate once what we need
+    // with no reallocation and moving around data.
+    // When allocating 16 branches, it only has to turn around and reallocate
+    // the right size and compact, which is less than ideal. In principle,
+    // with less allocations we should see less fragmentation in the arena.
+    std::uint16_t isBranch = 0;
```

Comment:
This probably isn't really much of a win, but if TaggedPointer goes to all its trouble, in context, it seems justified?